ReactSkia/sdk/NotificationCenter.h

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright (c) 2014 Sean Farrell
+// Copyright (C) 1994-2021 OpenTV, Inc. and Nagravision S.A.
+//  
+// Permission is hereby granted, free of charge, to any person obtaining a copy 
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights 
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
+// copies of the Software, and to permit persons to whom the Software is 
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in 
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
+// SOFTWARE.
+//
+
+#include <iostream>
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <list>
+#include <algorithm>
+
+#pragma once
+
+class NotificationCenter {
+    private:
+        struct ListenerBase {
+            ListenerBase() {}
+
+            ListenerBase(unsigned int i)
+            : id(i) {}
+
+            virtual ~ListenerBase() {}
+
+            unsigned int id;
+        };
+
+        template <typename... Args>
+        struct Listener : public ListenerBase {
+            Listener() {}
+
+            Listener(unsigned int i, std::function<void (Args...)> c)
+            : ListenerBase(i), cb(c) {}
+
+            std::function<void (Args...)> cb;
+        };
+
+        static std::mutex mutex;
+        static unsigned int last_listener;
+        static std::multimap<std::string, std::shared_ptr<ListenerBase>> listeners;
```

Comment:
How about design this PubSub as a common library and have an instance in RSkEventEmitter module?
Ideally, if the NotificationCenter be used in other places, the mutex should not share with RSkEventEmitter.

Reply:
Wanted to build NotificationCenter on the similar lines of https://developer.apple.com/documentation/foundation/nsnotificationcenter
The idea is allow the usage of observer pattern across the RN skia code.
For example: to implement TVEventHandler to provide Keyboard events to App
TVEventHandler (new module) -> inherits from RSkEventEmitter and implements virtual methods:
      - startObserving(): Registers for event RCTTVNavigationEventNotification using NotificationCenter
      - stopObserving(): Unregister RCTTVNavigationEventNotification using NotificationCenter
Window_unix -> Emits RCTTVNavigationEventNotification event using NotificationCenter for every keypress

Reply:
Maybe the second approach is to not have any static members in the NotificationCenter, and just create a global instance of NotificationCenter, named "defaultCenter" similar to what Apple implementation has and use that for TVEventHandler implementation.
In the similar way wherever PubSub is required, above pattern can be followed.

Reply:
well, in my opinion, would like to prevent global notification center as possible or it might be abused.
if that is really necessary, at least placing in platform implementations, e.g. in tv implementations.
considering this notification center is not be used by `RSkEventEmitter` and it also introduced 3rd party modified code. would recommend to remove this from ReactSkia/sdk. how do you think about this?

Reply:
As per your comments, and suggestions made by @munezbn, have removed the static definitions from Notification Center. Now, for each Set of notification individual Instances of this class can be created for managing specific PubSub needs.
However, a single (common) defaultCenter_ is created to manage default notifications.